src/basic/smack-util.c

```diff
@@ -45,84 +45,167 @@ bool mac_smack_use(void) {
+        assert(label);
 
-#ifdef HAVE_SMACK
         if (!mac_smack_use())
                 return 0;
 
-        if (label)
-                r = lsetxattr(path, "security.SMACK64", label, strlen(label), 0);
-        else
-                r = lremovexattr(path, "security.SMACK64");
-        if (r < 0)
+        s = getxattr(path, attr, l, NAME_MAX);
+        if (s < 0)
                 return -errno;
-#endif
 
-        return r;
+        ll = strdup(l);
+        if (!ll)
+                return -ENOMEM;
+
+        *label = ll;
+
+        return s;
+#else
+        return 0;
+#endif
 }
 
-int mac_smack_apply_fd(int fd, const char *label) {
-        int r = 0;
+static int mac_get_xattr_fd(int fd, const char *attr, const char **label) {
+#ifdef HAVE_SMACK
+        char l[NAME_MAX] = "", *ll = NULL;
+        ssize_t s = 0;
 
         assert(fd >= 0);
+        assert(attr);
+        assert(label);
 
-#ifdef HAVE_SMACK
         if (!mac_smack_use())
                 return 0;
 
-        if (label)
-                r = fsetxattr(fd, "security.SMACK64", label, strlen(label), 0);
-        else
-                r = fremovexattr(fd, "security.SMACK64");
-        if (r < 0)
+        s = fgetxattr(fd, attr, l, NAME_MAX);
+        if (s < 0)
                 return -errno;
-#endif
 
-        return r;
-}
+        ll = strdup(l);
+        if (!ll)
+                return -ENOMEM;
 
-int mac_smack_apply_ip_out_fd(int fd, const char *label) {
-        int r = 0;
+        *label = ll;
 
-        assert(fd >= 0);
+        return s;
+#else
+        return 0;
+#endif
+}
 
+static int mac_set_xattr(const char *path, const char *attr, const char *label) {
 #ifdef HAVE_SMACK
+        int r;
+
+        assert(path);
+        assert(attr);
+
         if (!mac_smack_use())
                 return 0;
 
         if (label)
-                r = fsetxattr(fd, "security.SMACK64IPOUT", label, strlen(label), 0);
+                r = lsetxattr(path, attr, label, strlen(label), 0);
         else
-                r = fremovexattr(fd, "security.SMACK64IPOUT");
+                r = lremovexattr(path, attr);
         if (r < 0)
                 return -errno;
+
 #endif
 
-        return r;
+        return 0;
 }
 
-int mac_smack_apply_ip_in_fd(int fd, const char *label) {
-        int r = 0;
+static int mac_set_xattr_fd(int fd, const char *attr, const char *label) {
+#ifdef HAVE_SMACK
+        int r;
 
         assert(fd >= 0);
+        assert(attr);
 
-#ifdef HAVE_SMACK
         if (!mac_smack_use())
                 return 0;
 
         if (label)
-                r = fsetxattr(fd, "security.SMACK64IPIN", label, strlen(label), 0);
+                r = fsetxattr(fd, attr, label, strlen(label), 0);
         else
-                r = fremovexattr(fd, "security.SMACK64IPIN");
+                r = fremovexattr(fd, attr);
         if (r < 0)
                 return -errno;
 #endif
 
-        return r;
+        return 0;
+}
+
+int mac_smack_read(const char *path, const char **label) {
```

Comment:
And as I said - the asserts need to be included here. Otherwise, you cannot deduct which function was called from the message that is printed.